File: 01046. Last Stone Weight/solution.js
/**
 * Problem: 1046. Last Stone Weight
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Returns the weight of the last stone after smashing the two heaviest repeatedly
 *
 * @param {number[]} stones - Stone weights
 *
 * @returns {number} - Last stone weight or 0
 */
// Define a function to  compute the last remaining stone's weight
const lastStoneWeight = (stones) => {
  // Continue looping while there is more than one stone left
  while (stones.length > 1) {
    // Sort the stones array in descending order to get the heaviest stones first
    const sortedStones = stones.sort((a, b) => b - a)

    // Smash the two heaviest stones: subtract the second heaviest from the heaviest
    sortedStones[1] = sortedStones[0] - sortedStones[1]
    // Remove the heaviest stone from the array after smashing
    sortedStones.shift()
  }

  // Return the weight of the last remaining stone, or undefined if none left
  return stones[0]
}

File: 01046. Last Stone Weight/solution.go
/**
 * Problem: 1046. Last Stone Weight
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

// MaxHeap defines a type for a max-heap of integers.
type MaxHeap []int

// Len returns the number of elements in the heap.
func (h MaxHeap) Len() int { return len(h) }

// Less compares two elements and returns true if the first is greater than the second (max-heap property).
func (h MaxHeap) Less(i, j int) bool { return h[i] > h[j] }

// Swap exchanges the elements at indices i and j.
func (h MaxHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }

// Push adds a new element to the heap.
func (h *MaxHeap) Push(x interface{}) { *h = append(*h, x.(int)) }

// Pop removes and returns the last element from the heap.
func (h *MaxHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1] // Get the last element.

	*h = old[:n-1] // Remove the last element.

	return x // Return the removed element.
}

// lastStoneWeight returns the weight of the last remaining stone after repeatedly smashing the two heaviest stones together.
func lastStoneWeight(stones []int) int {
	h := &MaxHeap{} // Initialize a new max-heap.
	heap.Init(h)    // Initialize the heap structure.

	for _, stone := range stones {
		heap.Push(h, stone) // Push each stone onto the heap.
	}

	for h.Len() > 1 {
		y := heap.Pop(h).(int) // Remove the heaviest stone.
		x := heap.Pop(h).(int) // Remove the second heaviest stone.

		if y != x {
			heap.Push(h, y-x) // If they are not equal, push the difference back onto the heap.
		}
	}

	if h.Len() == 0 {
		return 0 // If no stones are left, return 0.
	}

	return heap.Pop(h).(int) // Return the weight of the last remaining stone.
}
